jetty-servlet/src/test/java/org/eclipse/jetty/servlet/DispatcherTest.java

```diff
@@ -373,6 +377,98 @@ public void testForwardFilterToRogerServlet() throws Exception
         assertThat(rechoResponse, containsString("txeTohce"));
     }
 
+    @Test
+    public void testForwardCloseIntercepted() throws Exception
+    {
+        _contextHandler.addFilter(WrappingFilter.class, "/*", EnumSet.of(DispatcherType.REQUEST));
+        _contextHandler.addServlet(ForwardServlet.class, "/ForwardServlet/*");
+        _contextHandler.addServlet(AssertForwardServlet.class, "/AssertForwardServlet/*");
+
+        String expected =
+            "HTTP/1.1 200 OK\r\n" +
+                "Content-Type: text/html\r\n" +
+                "Content-Length: 0\r\n" +
+                "\r\n";
+
+        String responses = _connector.getResponse("GET /context/ForwardServlet?do=assertforward&do=more&test=1 HTTP/1.0\n\n");
+
+        assertEquals(expected, responses);
+    }
+
+    public static class WrappingFilter implements Filter
+    {
+        @Override
+        public void init(FilterConfig filterConfig) throws ServletException
+        {
+        }
+
+        @Override
+        public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException
+        {
+            ResponseWrapper wrapper = new ResponseWrapper((HttpServletResponse)response);
+            chain.doFilter(request, wrapper);
+            wrapper.sendResponse(response.getOutputStream());
+        }
+
+        @Override
+        public void destroy()
+        {
+        }
+    }
+
+    public static class ResponseWrapper extends HttpServletResponseWrapper
+    {
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+        public ResponseWrapper(HttpServletResponse response)
+        {
+            super(response);
+        }
+
+        @Override
+        public ServletOutputStream getOutputStream() throws IOException
+        {
+            return new ServletOutputStream()
+            {
+                @Override
+                public boolean isReady()
+                {
+                    return true;
+                }
+
+                @Override
+                public void setWriteListener(WriteListener writeListener)
+                {
+                    throw new UnsupportedOperationException();
+                }
+
+                @Override
+                public void write(int b) throws IOException
+                {
+                    buffer.write(b);
+                }
+
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException
+                {
+                    buffer.write(b, off, len);
+                }
+
+                @Override
+                public void close() throws IOException
+                {
+                    buffer.close();
```

Comment:
I have commented out this line `buffer.close()`, and the test still passes.

I'm not sure at this point what are we really testing? I would have expected the test to fail if `buffer.close()` was commented out?

If the point of the test is to verify that the wrapped `close()` is invoked, can you make it explicit with a `boolean` or a `CountDownLatch`?

Reply:
If you remove the method or call super.close() then the test fails.   The test correctly reproduces the failure produced by the client when the fix is not in place.